File: app/src/main/java/com/jiujiu/mvvmTemplate/ui/products/ProductsFragment.java
package com.jiujiu.mvvmTemplate.ui.products;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.recyclerview.widget.RecyclerView;

import com.jiujiu.mvvmTemplate.BR;
import com.jiujiu.mvvmTemplate.R;
import com.jiujiu.mvvmTemplate.databinding.FragmentProductsBinding;
import com.jiujiu.mvvmTemplate.ui.base.BaseFragment;


public class ProductsFragment extends BaseFragment<FragmentProductsBinding, ProductsFragViewModel> {
    private static final String TAG = "ProductsFragment";
    private static ProductsFragment mFragment;

    private ProductsRecyclerAdapter mAdapter;

    public ProductsFragment() {
        Log.d(TAG, "ProductsFragment: constructors");
    }

    public static ProductsFragment getInstance() {
        if (mFragment == null) {
            mFragment = new ProductsFragment();
        }
        return mFragment;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupLayout();
        setupViewModel();
        Log.d(TAG, "onViewCreated: ");
    }

    private void setupLayout() {
        mAdapter = new ProductsRecyclerAdapter();
        getBinding().rvProducts.setAdapter(mAdapter);
        getBinding().rvProducts.addOnItemTouchListener(new RecyclerView.SimpleOnItemTouchListener() {
        });
    }

    private void setupViewModel() {
        getViewModel().getProductLiveData().observe(this, products -> this.mAdapter.setmData(products));
    }

    @Override
    protected int getBindingVariableId() {
        return BR.viewModel;
    }

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_products;
    }

    @Override
    protected Class<ProductsFragViewModel> getViewModelType() {
        return ProductsFragViewModel.class;
    }

}

File: app/src/main/java/com/jiujiu/mvvmTemplate/ui/products/ProductsFragment.kt
package com.jiujiu.mvvmTemplate.ui.products

import android.os.Bundle
import android.view.View
import androidx.lifecycle.Observer
import com.jiujiu.mvvmTemplate.BR
import com.jiujiu.mvvmTemplate.R
import com.jiujiu.mvvmTemplate.databinding.FragmentProductsBinding
import com.jiujiu.mvvmTemplate.ui.base.BaseFragment
import org.jetbrains.anko.info


class ProductsFragment : BaseFragment<FragmentProductsBinding, ProductsFragViewModel>() {

    private lateinit var mAdapter: ProductsRecyclerListerAdapter

    override val bindingVariableId: Int
        get() = BR.viewModel

    override val layoutId: Int
        get() = R.layout.fragment_products

    override val viewModelType: Class<ProductsFragViewModel>
        get() = ProductsFragViewModel::class.java


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        info { "onCreated" }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupLayout()
        setupViewModel()
        info("onViewCreated: ")
    }

    override fun onDestroyView() {
        super.onDestroyView()
        info("onDestroyView:")
    }

    override fun onDestroy() {
        super.onDestroy()
        info("onDestroy")
    }

    private fun setupLayout() {
        mAdapter = ProductsRecyclerListerAdapter()
        binding.rvProducts.adapter = mAdapter
    }

    private fun setupViewModel() {
        viewModel.productLiveData.observe(this, Observer { mAdapter.submitList(it) })
    }

    companion object {
        fun getInstance(): ProductsFragment = ProductsFragment()
    }

}
